veracruz-mcu-client/vc.c

```diff
@@ -0,0 +1,577 @@
+        size_t len;
+        err = mbedtls_asn1_get_len(&ext_seq_ptr, ext_seq_end, &len);
+        if (err) {
+            printf("asn1 parsing failed (%d)\n", err);
+            return err;
+        }
+
+        if (tag == (MBEDTLS_ASN1_SEQUENCE | MBEDTLS_ASN1_CONSTRUCTED)) {
+            bool runtime_hash_found = false;
+
+            // search for our oid
+            uint8_t *ext_ptr = ext_seq_ptr;
+            uint8_t *ext_end = ext_ptr + len;
+            while (ext_ptr < ext_end) {
+                uint8_t tag = *ext_ptr++;
+                size_t len;
+                err = mbedtls_asn1_get_len(&ext_ptr, ext_end, &len);
+                if (err) {
+                    printf("asn1 parsing failed (%d)\n", err);
+                    return err;
+                }
+
+                if (tag == MBEDTLS_ASN1_OID) {
+                    if (memcmp(ext_ptr, VC_RUNTIME_HASH_EXTENSION_ID,
+                            sizeof(VC_RUNTIME_HASH_EXTENSION_ID)) == 0) {
+                        runtime_hash_found = true;
+                    }
+                    break;
+                }
+
+                ext_ptr += len;
+            }
+
+            // found oid for runtime hash, lookup actual hash value
+            if (runtime_hash_found) {
+                // search for runtime hash
+                uint8_t *ext_ptr = ext_seq_ptr;
+                uint8_t *ext_end = ext_ptr + len;
+                while (ext_ptr < ext_end) {
+                    uint8_t tag = *ext_ptr++;
+                    size_t len;
+                    err = mbedtls_asn1_get_len(&ext_ptr, ext_end, &len);
+                    if (err) {
+                        printf("asn1 parsing failed (%d)\n", err);
+                        return err;
+                    }
+
+                    if (tag == MBEDTLS_ASN1_OCTET_STRING) {
+                        // finally found our octet string, does it match?
+                        printf("found: ");
+                        hex(ext_ptr, len);
+                        printf("\n");
+
+                        // check against known runtime hashes
+                        if (len == 32) {
+                            for (int i = 0; i < sizeof(VC_RUNTIME_HASHES)/32; i++) {
+                                if (memcmp(ext_ptr, VC_RUNTIME_HASHES[i], 32) == 0) {
+                                    printf("\033[32mverified runtime hash:\033[m ");
+                                    hex(ext_ptr, len);
+                                    printf("\n");
+                                    return 0;
+                                }
+                            }
+                        }
+
+                        printf("runtime hash mismatch\n");
+                        return -EBADE;
+                    }
+
+                    ext_ptr += len;
+                }
+            }
+        }
+
+        ext_seq_ptr += len;
+    }
+
+    printf("no runtime hash?\n");
+    return -EBADE;
+}
+
+int vc_connect(vc_t *vc) {
```

Comment:
I second @dreemkiller's call for more comments explaining what some of the functions in this file are doing.  Just as we've adopted `rustdoc` style comments for Rust code, could we adopt `doxygen`-style comments for the C code properly documenting what is going on?